NanoProd/plugins/TauSpinnerTableProducer.cc

```diff
@@ -0,0 +1,246 @@
+    return TauSpinner::SimpleParticle(
+        input_part.px(), input_part.py(), input_part.pz(), input_part.energy(), input_part.pdgId());
+  }
+
+  std::vector<std::pair<std::string, double>> nameAndValue(const std::vector<double> &val_vec) const {
+    std::vector<std::pair<std::string, double>> out;
+    for (auto val : val_vec) {
+      std::string name = std::to_string(val);
+      name.erase(name.find_last_not_of('0') + 1, std::string::npos);
+      name.erase(name.find_last_not_of('.') + 1, std::string::npos);
+      size_t pos = name.find(".");
+      if (pos != std::string::npos)
+        name.replace(pos, 1, "p");
+      pos = name.find("-");
+      if (pos != std::string::npos)
+        name.replace(pos, 1, "minus");
+      out.push_back(std::make_pair(name, val));
+    }
+    return out;
+  }
+
+  void printModuleInfo(edm::ParameterSet const &config) const {
+    std::string header = (boost::format("%-39s%39s") % config.getParameter<std::string>("@module_type") %
+                          config.getParameter<std::string>("@module_label"))
+                             .str();
+    std::cout << std::string(78, '-') << "\n";
+    std::cout << header << "\n";
+    // std::cout << "Produces: " << flow << "\n";
+    std::cout << boost::format("%-15s : %-60s\n") % "Input" % config.getParameter<edm::InputTag>("input").encode();
+    std::cout << boost::format("%-15s : %-60s\n") % "Branch" % config.getParameter<std::string>("branch");
+    std::string thetaStr;
+    for (const auto &theta : theta_vec_)
+      thetaStr += theta.first + ",";
+    std::cout << boost::format("%-15s : %-60s\n") % "Theta" % thetaStr;
+  }
+
+  edm::EDGetTokenT<edm::View<reco::GenParticle>> genPartsToken_;
+  std::string branch_;
+  std::vector<std::pair<std::string, double>> theta_vec_;
+  int bosonPdgId_;
+  std::string tauSpinnerPDF_;
+  bool ipp_;
+  int ipol_;
+  int nonSM2_;
+  int nonSMN_;
+  double cmsE_;
+};
+
+TauSpinnerTableProducer::TauSpinnerTableProducer(const edm::ParameterSet &config)
+    : genPartsToken_(consumes(config.getParameter<edm::InputTag>("input"))),
+      branch_(config.getParameter<std::string>("branch")),
+      bosonPdgId_(25),                                             //Higgs, to be configurable?
+      tauSpinnerPDF_(config.getParameter<std::string>("pdfSet")),  //"NNPDF30_nlo_as_0118"
+      ipp_(true),                                                  //pp collisions
+      ipol_(0),
+      nonSM2_(0),
+      nonSMN_(0),
+      cmsE_(config.getParameter<double>("cmsE"))  //cms energy in GeV, 13000.0
+{
+  theta_vec_ = nameAndValue(config.getParameter<std::vector<double>>("theta"));
+
+  printModuleInfo(config);
+
+  //state that we use tauola/tauspinner resource
+  usesResource(edm::SharedResourceNames::kTauola);
+
+  produces<nanoaod::FlatTable>();
+}
+
+reco::GenParticle TauSpinnerTableProducer::getBoson(const edm::View<reco::GenParticle> &parts, bool &foundBoson) const {
```

Comment:
better to return a `std::set<const reco::GenParticle*>` and check that set size is 1. Otherwise, it could find a boson  for `HH` or `HHH` samples and it could be `H->tautau`.

Reply:
Yes, as it is the module returns always 1st found lastCopy of a particle with a given pdgID (here Higgs) if it is found. Without `break` in the search loop one can check if the found boson is unique. And yes, when a container is returned one can check if it has unique item to be further processed or is empty or has more then one item. I will follow it up.